core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,118 @@
+
+    public open fun isEmpty(): Boolean = intProgression.isEmpty()
+
+    override fun toString(): String = if (intProgression.step > 0) "$first..$last step ${intProgression.step}" else "$first downTo $last step ${intProgression.step}"
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
+
+        other as LocalDateProgression
+
+        return intProgression == other.intProgression
+    }
+
+    override fun hashCode(): Int {
+        return intProgression.hashCode()
+    }
+
+
+    public companion object {
+        public fun fromClosedRange(
+            rangeStart: LocalDate,
+            rangeEnd: LocalDate,
+            stepValue: Int,
+            stepUnit: DateTimeUnit.DayBased
+        ): LocalDateProgression = LocalDateProgression(rangeStart, rangeEnd, stepValue * stepUnit.days)
+    }
+}
+
+public class LocalDateRange(start: LocalDate, endInclusive: LocalDate) : LocalDateProgression(start, endInclusive, 1), ClosedRange<LocalDate>, OpenEndRange<LocalDate> {
+    override val start: LocalDate get() = first
+    override val endInclusive: LocalDate get() = last
+    override val endExclusive: LocalDate get() = endInclusive.plus(1, DateTimeUnit.DAY)
+
+    @Suppress("ConvertTwoComparisonsToRangeCheck")
+    override fun contains(value: LocalDate): Boolean = first <= value && value <= last
+
+    override fun isEmpty(): Boolean = first > last
+
+    override fun toString(): String = "$first..$last"
+
+    public companion object {
+        private val DATE_ONE = LocalDate(1, 1, 1)
+        private val DATE_TWO = LocalDate(1, 1, 2)
+        public val EMPTY: LocalDateRange = LocalDateRange(DATE_TWO, DATE_ONE)
+    }
+}
+
+public fun LocalDateProgression.first(): LocalDate {
+    if (isEmpty())
+        throw NoSuchElementException("Progression $this is empty.")
+    return this.first
+}
+public fun LocalDateProgression.last(): LocalDate {
+    if (isEmpty())
+        throw NoSuchElementException("Progression $this is empty.")
+    return this.last
+}
+public fun LocalDateProgression.firstOrNull(): LocalDate? = if (isEmpty()) null else this.first
+public fun LocalDateProgression.lastOrNull(): LocalDate? = if (isEmpty()) null else this.last
+
+public fun LocalDateProgression.reversed(): LocalDateProgression = LocalDateProgression(intProgression.reversed())
+
+public fun LocalDateProgression.step(value: Int, unit: DateTimeUnit.DayBased) : LocalDateProgression = LocalDateProgression(intProgression.step(value * unit.days))
+public fun LocalDateProgression.step(value: Long, unit: DateTimeUnit.DayBased) : LocalDateProgression = step(value.toInt(), unit)
+
+public infix fun LocalDate.downTo(that: LocalDate) : LocalDateProgression = LocalDateProgression.fromClosedRange(this, that, -1, DateTimeUnit.DAY)
+public infix fun LocalDate.downUntil(that: LocalDate) : LocalDateProgression = downTo(that.plus(1, DateTimeUnit.DAY))
```

Comment:
I don't think we have a `downUntil` function for any other ranges. Consistency of experience is extremely important to us: we want our users' knowledge to be completely transferable across the API entries we provide. "Aha, I need an `Int` range, and I already know how to do this for `LocalDate` ranges" should always work, unless there's a significant barrier to that.

Also, overflow.

Reply:
I added it for symmetry with `rangeUntil`, but I can remove it.